File: bin/satest/bench/2.py
import time
import statistics

def run_tests():
    # Registers
    r1 = 0          # Counter
    r2 = 1          # Increment
    r8 = 1_000_000  # Stop value

    while True:
        # vadd r1 = r1 + r2 (+ offset 1, simulated)
        r1 = r1 + r2 + 1  # the assembler offset

        # #eq macro: compare r1 and r8
        r7 = 1 if r1 == r8 else 0

        # jif (jump if non-zero)
        if r7 != 0:
            break  # jump to mark 2

        # else continue looping (jmp 1)
        # loop continues automatically


# Execution loop
latencies = []

for _ in range(100):
    start = time.perf_counter_ns()
    run_tests()
    end = time.perf_counter_ns()
    latencies.append(end - start)

median_latency = statistics.median(latencies)

print(f"Median Execution Time: {median_latency:.2f} ns")

File: bin/satest/bench/2.js
function runTests() {
  let ctr = 0;

  while (ctr < 1000000) {
    ctr += 1;
  }
}

const iterations = 100;
const latencies = new BigUint64Array(iterations);

// Warm up - Let V8 optimize the function
for (let i = 0; i < 200_000; i++) runTests();

for (let i = 0; i < iterations; i++) {
  const start = process.hrtime.bigint();
  runTests();
  const end = process.hrtime.bigint();
  latencies[i] = end - start;
}

// Calculate Median
latencies.sort();
const median = latencies[iterations / 2];

console.log(`Median Execution Time: ${median} ns`);
